snuba/migrations/migration.py

```diff
@@ -0,0 +1,65 @@
+from abc import ABC, abstractmethod, abstractproperty
+from typing import Optional, Sequence
+
+from snuba.migrations.operations import Operation
+
+
+class Migration(ABC):
+    """
+    A migration consists of one or more forward operations which will be executed
+    on all of the local and distributed nodes of the cluster. Upon error, the backwards
+    methods will be executed. The backwards operations are responsible for returning
+    the system to its pre-migration state, so that the forwards methods can be safely
+    retried.
+
+    Once the migration has been completed, we shouldn't use the backwards methods
+    to try and go back to the prior state. Since migrations can delete data, attempting
+    to revert cannot always bring back the previous state completely.
+
+    The operations in a migration should bring the system from one consistent state to
+    the next. There isn't a hard and fast rule about when operations should be grouped
+    into a single migration vs having multiple migrations with a single operation
+    each. Generally if the intermediate state between operations is not considered to
+    be valid, they should be put into the same migration. If the operations are
+    completely unrelated, they are probably better as separate migrations.
+
+    Migrations that cannot be completed immediately, such as those that contain
+    a data migration, must be marked with is_dangerous = True.
+
+    The easiest way to run dangerous migrations will be with downtime. If Snuba is not
```

Comment:
I would not call the blocking migrations `dangerous`. This term does not convery the meaning that this migration requires manual intervention and prevents you from skipping the version. Moreover in sentry the same word is using for something different, so it would be confusing.
What about calling them `manual` or `blocking` ?

Reply:
I wouldn't use the word `manual`, because it's not necessarily manual unless you need to run without downtime. I'll update to use `blocking` instead.